lib/source.js

```diff
@@ -47,50 +47,116 @@ module.exports = class CovSource {
     }
 
-    if (!(start.source && end.source)) {
-      return {}
-    }
+    const sourceMapIterator = sourceMapScanner.getIterator()
+    let start = sourceMapIterator.scanTo(lines[0].line, startCol - lines[0].startCol)
+    let last = start
+    let endPos = { line: lines[lines.length - 1].line, col: endCol - lines[lines.length - 1].startCol }
+    let returnLocs = []
+    while (true) {
+      let next = sourceMapIterator.next()
+      const isPastEnd = !next || sourceMapIterator.isGeneratedAfter(next, endPos.line, endPos.col)
+
+      if (isPastEnd || next.source !== start.source ||
+        next.originalLine < start.originalLine ||
+        (next.originalLine === start.originalLine && next.originalColumn < start.originalColumn)) {
+        if (last !== start) {
+          returnLocs.push({
+            sourceFile: start.source,
+            startLine: start.originalLine,
+            relStartCol: start.originalColumn,
+            endLine: last.originalLine,
+            relEndCol: last.originalColumn
+          })
+        }
+        if (isPastEnd) {
+          break
+        }
+        start = next
+      } else {
 
-    if (start.source !== end.source) {
-      return {}
-    }
+        // now we now we aren't going past the start (or different source) or the end etc.
+        // if the next token is going back past the last, ignore it.
+        if (next.originalLine < last.originalLine ||
+          (next.originalLine === last.originalLine && next.originalColumn < last.originalColumn)) {
+          continue
+        }
+      }
 
-    if (start.line === end.line && start.column === end.column) {
-      end = sourceMap.originalPositionFor({
-        line: lines[lines.length - 1].line,
-        column: endCol - lines[lines.length - 1].startCol,
-        bias: LEAST_UPPER_BOUND
-      })
-      end.column -= 1
+      last = next
     }
+    return this._removeOverlapping(returnLocs)
+  }
 
-    return {
-      startLine: start.line,
-      relStartCol: start.column,
-      endLine: end.line,
-      relEndCol: end.column
+  // minimize overlapping ranges to avoid bumping up the branches/functions count unless very necessary
+  _removeOverlapping (returnLocs) {
+    if (returnLocs.length <= 1) {
+      return returnLocs
     }
+    const bySource = Object.create(null)
+    returnLocs.forEach((loc) => {
+      if (!bySource[loc.source]) {
+        bySource[loc.source] = []
+      }
+      bySource[loc.source].push(loc)
+    })
+
+    const dedupedLocs = []
+    Object.keys(bySource).forEach((source) => {
+      let locs = bySource[source]
+      while (locs.length) {
+        const loc1 = locs.pop()
+        let isOverlapped = false
+
+        for (let i = 0; i < locs.length; i++) {
+          const loc2 = locs[i]
+          if (loc2 === loc1) {
+            return
+          }
+
+          let isEndLoc1BeforeLoc2Start = ((loc1.endLine < loc2.startLine) || (loc1.endLine === loc2.startLine && loc1.relEndCol < loc2.relStartCol))
```

Comment:
Shouldn't the SourceMap generator be making an effort to avoid overlapping source files? I wonder if we could add this logic later.